src/solver/goldensectionsearch/mod.rs

```diff
@@ -0,0 +1,139 @@
+pub struct GoldenSectionSearch<O: ArgminOp> {
+    min_bound: f64,
+    max_bound: f64,
+    init_estimate: f64,
+    tolerance: f64,
+
+    x0: f64,
+    x1: f64,
+    x2: f64,
+    x3: f64,
+    f1: O::Output,
+    f2: O::Output,
+}
+
+impl<O> GoldenSectionSearch<O>
+where
+    O: ArgminOp,
+    O::Output: ArgminZero,
+{
+    /// Constructor
+    pub fn new(min_bound: f64, max_bound: f64) -> Self {
+        GoldenSectionSearch {
+            min_bound,
+            max_bound,
+            init_estimate: 0.0,
+            tolerance: 0.01,
+            x0: min_bound,
+            x1: 0.0,
+            x2: 0.0,
+            x3: max_bound,
+            f1: O::Output::zero(),
+            f2: O::Output::zero(),
+        }
+    }
+
+    /// Set tolerance
+    pub fn tolerance(mut self, tol: f64) -> Self {
+        self.tolerance = tol;
+        self
+    }
+}
+
+impl<O> Solver<O> for GoldenSectionSearch<O>
+where
+    O: ArgminOp<Param = f64, Output = f64>,
+{
+    const NAME: &'static str = "Golden-section search";
+
+    fn init(
+        &mut self,
+        op: &mut OpWrapper<O>,
+        state: &IterState<O>,
+    ) -> Result<Option<ArgminIterData<O>>, Error> {
+        let init_estimate = state.param;
+        if init_estimate < self.min_bound || init_estimate > self.max_bound {
+            Err(ArgminError::InvalidParameter {
+                text: "Initial estimate must be ∈ [min_bound, max_bound].".to_string(),
+            }
+            .into())
+        } else {
+            let ie_min = init_estimate - self.min_bound;
+            let max_ie = self.max_bound - init_estimate;
+            let (x1, x2) = if max_ie.abs() > ie_min.abs() {
+                (init_estimate, init_estimate + G2 * max_ie)
+            } else {
+                (init_estimate - G2 * ie_min, init_estimate)
+            };
+            self.x1 = x1;
+            self.x2 = x2;
+            self.f1 = op.apply(&self.x1)?;
+            self.f2 = op.apply(&self.x2)?;
+            Ok(Some(ArgminIterData::new()))
+        }
+    }
+
+    fn next_iter(
+        &mut self,
+        op: &mut OpWrapper<O>,
+        _state: &IterState<O>,
+    ) -> Result<ArgminIterData<O>, Error> {
+        if self.tolerance * (self.x1.abs() + self.x2.abs()) >= (self.x3 - self.x0).abs() {
+            return Ok(ArgminIterData::new()
+                .param(if self.f1 < self.f2 { self.x1 } else { self.x2 })
+                .termination_reason(TerminationReason::TargetToleranceReached));
+        }
+
+        if self.f2 < self.f1 {
+            self.x0 = self.x1;
+            self.x1 = self.x2;
+            self.x2 = G1 * self.x1 + G2 * self.x3;
+            self.f1 = self.f2;
+            self.f2 = op.apply(&self.x2)?;
+        } else {
+            self.x3 = self.x2;
+            self.x2 = self.x1;
+            self.x1 = G1 * self.x2 + G2 * self.x0;
+            self.f2 = self.f1;
+            self.f1 = op.apply(&self.x1)?;
+        }
+        Ok(ArgminIterData::new())
```

Comment:
If I understand this correctly, the smaller one of `self.f1` and `self.f2` is your current best cost function value and `self.x1` or `self.x2` is the corresponding parameter vector. If you could add them here to the `ArgminIterData` struct as well (via `param()` and `cost()`), then they will be available to observers. 

The same applies to line 119 regarding `cost()`.

You can then access these values in the `state` variable passed to `next_iter` via `get_param()` and `get_cost()`, which would spare you the check in line 120. But this is a matter of taste.